pkg/sql/catalog/funcdesc/func_desc.go

```diff
@@ -836,14 +836,36 @@ func IsOIDUserDefinedFunc(oid oid.Oid) bool {
 	return catid.IsOIDUserDefined(oid)
 }
 
-// CheckLeakProofVolatility returns an error when a function is defined as
+// ValidateLeakProofVolatility returns an error when a function is defined as
 // leakproof but not immutable. See more details in comments for volatility.V.
-func CheckLeakProofVolatility(fn catalog.FunctionDescriptor) error {
-	if fn.GetLeakProof() && fn.GetVolatility() != catpb.Function_IMMUTABLE {
-		return pgerror.Newf(
-			pgcode.InvalidFunctionDefinition,
+func ValidateLeakProofVolatility(
+	options tree.FunctionOptions, curLeakProof bool, curVolatility catpb.Function_Volatility,
+) error {
+	for _, option := range options {
+		switch t := option.(type) {
+		case tree.FunctionVolatility:
+			v, err := VolatilityToProto(t)
+			if err != nil {
+				panic(err)
+			}
+			curVolatility = v
+		case tree.FunctionLeakproof:
+			curLeakProof = bool(t)
+		}
+	}
+
+	if err := mustMatchLeakProofVolatility(curLeakProof, curVolatility); err != nil {
+		return pgerror.Wrap(err, pgcode.InvalidFunctionDefinition, "invalid volatility")
```

Comment:
I think my preference would be to have this return an unwrapped error and then wrap the pgcode above.

Reply:
Make sense. It is weird returning pg errors from this package.